File: clojure/nucleotide-count/_iterations/1/src/nucleotide_count.clj
(ns nucleotide-count)

(def valid-nucleotide? #{\A \C \G \T})

(defn- validate-nucleotide [nucleotide]
  (when-not (valid-nucleotide? nucleotide)
    (throw (IllegalArgumentException. (str "Invalid nucleotide: " nucleotide)))))

(defn- validate-strand [strand]
  (doseq [nucleotide strand]
    (validate-nucleotide nucleotide)))

(defn count-of-nucleotide-in-strand [nucleotide strand]
  (validate-nucleotide nucleotide)
  (validate-strand strand)
  (count (filter #(= nucleotide %) strand)))

(defn nucleotide-counts [strand]
  (validate-strand strand)
  (->> (frequencies strand)
       (merge {\A 0 \C 0 \G 0 \T 0})))

File: julia/nucleotide-count/_iterations/1/nucleotide-count.jl
"""
    count_nucleotides(strand)

The count of each nucleotide within `strand` as a dictionary.

Invalid strands raise a `DomainError`.

"""
function count_nucleotides(strand)
    if occursin(r"[^ACGT]", strand)
        throw(DomainError(strand, "Invalid DNA strand: $strand"))
    end
    
    counts = Dict(n => 0 for n in "ACGT")
    for n in strand
        counts[n] += 1
    end
    counts
end
